python/tank/bootstrap/manager.py

```diff
@@ -236,32 +244,153 @@ def bootstrap_engine(self, engine_name, entity=None):
         the engine may not be the same as the API version that was
         executed during the bootstrap.
 
-        :param entity: Shotgun entity to launch engine for
-        :type entity: Dictionary with keys type and id
-        :param engine_name: name of engine to launch (e.g. ``tk-nuke``)
-        :returns: :class:`sgtk.platform.Engine` instance
+        :param engine_name: Name of engine to launch (e.g. ``tk-nuke``).
+        :param entity: Shotgun entity to launch engine for.
+        :returns: :class:`sgtk.platform.Engine` instance.
         """
-        log.info("Bootstrapping into engine %s for entity %s." % (engine_name, entity))
 
-        log.debug("Bootstrapping into environment.")
+        log.info("Bootstrapping engine %s for entity %s." % (engine_name, entity))
+
         tk = self._bootstrap_sgtk(engine_name, entity)
 
-        log.debug("Resolving context.")
-        if entity is None:
-            ctx = tk.context_empty()
+        engine = self._start_engine(tk, engine_name, entity)
+
+        return engine
+
+    def bootstrap_engine_async(self,
+                               engine_name,
+                               entity=None,
+                               progress_callback=None,
+                               completed_callback=None,
+                               failed_callback=None):
+        """
+        Create an :class:`sgtk.Sgtk` instance for the given engine and entity,
+        then launch into the given engine.
+
+        The :class:`sgtk.Sgtk` instance will be bootstrapped asynchronously in a background thread,
+        followed by launching the engine synchronously in the main application thread.
+        This will allow the main application to continue its execution and
+        remain responsive when bootstrapping the toolkit involves downloading files and
+        installing apps from the toolkit app store.
+
+        If entity is None, the method will bootstrap into the site
+        config. This method will attempt to resolve the config according
+        to business logic set in the associated resolver class and based
+        on this launch a configuration. This may involve downloading new
+        apps from the toolkit app store and installing files on disk.
+
+        3 callback functions can be provided.
+
+        A callback function that reports back on the bootstrap progress
+        with the following signature::
+
+            progress_callback(progress_value, message, current_index, maximum_index)
```

Comment:
Done, including some magic to handle the following callback chain when bootstrapping asynchronously:

> ToolkitManager._report_progress()
> --> BootstrapWorker._report_progress()
> --> signal BootstrapWorker.progressing
> --> slot BootstrapSupervisor._progress_bootstrap()
> --> client progress callback